keps/sig-api-machinery/20200309-consistent-resource-versions-semantics.md

```diff
@@ -0,0 +1,216 @@
+
+I am currently considering not including `ResourceVersionMatch` since:
+- The header approach sets precidence for an approach I'm not sure we want to encourage in api-machinery
+- get currently does not have the semantic consistency problems of list, and so does not urgently need this parameter
+- it is possible to use list to get a single item at a specific resource version already
+
+### Risks and Mitigations
+
+The main risk to this proposal that it complicates the API surface area, resulting in an API that is
+more difficult to understand and use. But the existing behavior already complicates the API. With
+this proposal, clients at least have the ability to use a consistent set of semantics.
+
+Another risk is that clients will either not realize, or not be sufficiently
+motivated, to update their code to move away from the legacy behavior. This
+can be mitigated a couple ways:
+- Update client bindings (client-go, ...) to discourage using the
+  legacy behavior, and eventually to disallow it.
+- At some point in the future, start logging warning on the server when the
+  legacy behavior is used to make it more obvious what needs to be changed?
+
+## Alternatives Considered
+
+### Alternative: Introduce ExactResourceVersion and MinResourceVersion parameters
+
+Deprecate `ResourceVersion` and introduce `ExactResourceVersion` and `MinResourceVersion`.
+
+The three cases are equivalent to the `ResourceVersionMatch` cases from option 1
+and would use the equivalent documentation.
+
+This makes it obvious that the `ResourceVersion` parameter is deprecated. It
+does this as the API aesthetic cost of having a top level parameter be forever
+deprecated.
+
+**Backward Compatibility:**
+
+Versions of the kube-apiserver that pre-date the introduction of the `ExactResourceVersion`
+and `MinResourceVersion` parameters will ignore them, resulting in a quorum read. For clients
+to tolate responses from servers that ignore the new parameters in a backward compatible way:
+
+- When using `MinResourceVersion`, clients can either:
+  - Also include `ResourceVersion` and get legacy semantics if `MinResourceVersion` is ignored.
+    They will need to handle HTTP 410 “Gone” responses if Limit is set.
+  - Tolerate a quorum read, which is guaranteed to provided the newest ResourceVersion
+    but has scalability/performance implications.
+- When using `ExactResourceVersion` also set `ResourceVersion` and check the ResourceVersion in the ListMeta of the response.
+  If it does not match, fall back to a request with a Limit set.
+
+**Advantages**
+
+- Deprecating `ResourceVersion` highlights to API users that they need change calls that use ResourceVersion.
+
+**Disadvantages**
+
+- Clients have to include `ResourceVersion` even when using the new parameters for backward compatibility.
+- `ResourceVersion` becomes deperacated but can never be removed.
```

Comment:
I think that's the main disadvantage. In theory it's deprecated, but in practice we have 3 options to understand (instead of two).

Reply:
+1, clarifying further